### ISSUE TITLE
onRetry obj.attempt undefined when intelligent uploading is enabled

### ISSUE THREAD

**Issue by wuservices** [filestack/filestack-js#72]
On 0.9.12, In my `uploadConfig`, I have
```
 onRetry: (obj) ->
    console.debug "Retrying #{obj.location} (for #{obj.filename}, attempt #{obj.attempt} of 10" +
    ", location #{obj.location}.", obj
```

However, once my account got intelligent uploading enabled on it, I noticed that `obj.attempt` was undefined and simply didn't exist in `obj` above. I verified this hypothesis by setting `intelligent: false`, and the attempt number came back.

**Comment by velveteer** [comment]
@wuservices Sorry this is poorly documented, but attempts are not counted in the intelligent flow because attempts are replaced with the `chunkSize` metric. We can patch this to include `attempt` on the retry object for intelligent uploads, but it will always be 0.

**Comment by wuservices** [comment]
@velveteer thanks for explaining that. Figured maybe it wasn't as relevant or something.

Sorry to get a little off topic here, but I'm hoping you can be a huge help here. I've been trying to chase down a lot of customer complaints we've been having about failed uploads, so I'm trying to figure out how both standard flow and intelligent flow work, to try and fix these issues.

My current test case is uploading an 18 MB image, while setting Chrome's Network throttling to Fast 3G (0.5 mbps upload). This sadly happens to match many consumers' DSL connections. Without intelligent uploading on, this was guaranteed to fail every time since it would upload 3, 6 MB chunks in parallel, and would always time out.

Let's say you can upload at about approx 50 kb / sec on a 3G or slow DSL connection. Best case, it would take 2 min to upload a single chunk if you could use the entire connection, and it would take 6 minutes to upload all 3 chunks in parallel. I think this explains why non-intelligent upload would never work for us.

_Does this hypothesis / analysis make sense so far?_

Now, with intelligent ingestion on, I'm trying this all again. I have a ticket open to get briefed on how exactly it works technically, but here's my best guess on how it works.  You chunk the files out into normal uploads yourself, since multipart direct to S3 has a 5 MB min chunk size. Then, you combine them with a commit with your origin servers once all the parts are done. With II, you start at 8 MB chunks and each time a chunk fails or times out, you cut the chunk size in half and retry, and you continue to reduce the chunk size until things succeed. You make no attempts to determine the connection speed to see if you should be reducing the chunk size faster than a factor of 2 (retryFactor?) on a retry, in a scenario where the connection is super slow.

_Is this loosely how intelligent ingestion works?_

If so, I'd expect that on my 18 MB file test on a 50 kb/sec upload connection, and the default settings (2 min timeout, 2 retryFactor, 3 concurrency), nothing would succeed until each chunk was under approx 2 MB since 50 kb/sec / 3 concurrents * 120 sec would be 2000 kb.

In my test, that's basically exactly what happens, and it's pretty inefficient, since it just keeps starting over again. The end result takes almost 12 minutes to upload for an effective rate of about 25 kb/sec (half the actual connection speed).

```
12:14:46 Start upload
12:15:56 Retrying upload part 1 (for IMG_5245.jpg, attempt undefined of 10, location upload part 1. {location: "upload part 1", parts: Array(3), filename: "IMG_5245.jpg", chunkSize: 4194304}
12:15:57 Retrying upload part 3 (for IMG_5245.jpg, attempt undefined of 10, location upload part 3. {location: "upload part 3", parts: Array(3), filename: "IMG_5245.jpg", chunkSize: 4194304}
12:17:29 Retrying upload part 2 (for IMG_5245.jpg, attempt undefined of 10, location upload part 2. {location: "upload part 2", parts: Array(3), filename: "IMG_5245.jpg", chunkSize: 4194304}
12:17:58 Retrying upload part 1 (for IMG_5245.jpg, attempt undefined of 10, location upload part 1. {location: "upload part 1", parts: Array(3), filename: "IMG_5245.jpg", chunkSize: 2097152}
12:19:32 Retrying upload part 2 (for IMG_5245.jpg, attempt undefined of 10, location upload part 2. {location: "upload part 2", parts: Array(3), filename: "IMG_5245.jpg", chunkSize: 2097152}
12:20:03 Retrying upload part 1 (for IMG_5245.jpg, attempt undefined of 10, location upload part 1. {location: "upload part 1", parts: Array(3), filename: "IMG_5245.jpg", chunkSize: 1048576}
12:21:35 Retrying upload part 2 (for IMG_5245.jpg, attempt undefined of 10, location upload part 2. {location: "upload part 2", parts: Array(3), filename: "IMG_5245.jpg", chunkSize: 1048576}
12:22:07 Retrying upload part 1 (for IMG_5245.jpg, attempt undefined of 10, location upload part 1. {location: "upload part 1", parts: Array(3), filename: "IMG_5245.jpg", chunkSize: 524288}
```

_Is the best solution just to increase the timeout by a lot? Any other suggestions?_

_Will the uploader ever try to adjust chunk size based on connection speed?_

**Comment by velveteer** [comment]
@wuservices Great observations all around.

I believe the default timeout is too low and that is why failures are occurring when they should not. FII is usually reserved for networks where some QoS may be interfering, especially on mobile networks.

We will probably look into using connection speed to help determine chunk size, and increasing the default timeout or simply removing the timeout from the S3 PUT may land in a future patch release. Right now FII chunk size is only calculated on network failure or timeout.

There is an alternative value for the `intelligent` option named `'fallback'`, this will use the regular upload flow up until there is failure in which case it switches to FII. Or you could continue using FII with a smaller `intelligentChunkSize` as well -- so there are a few options around FII that could help with overall speed. Outside of FII your best bet is increasing the timeout value.

**Comment by wuservices** [comment]
@velveteer thanks for checking this out. I agree the default timeout is way too low. There are plenty of people stuck on connections with a few hundred kbps upload, or on public Wi-Fi, where there's QoS like you mentioned. It seems like the intelligent ingestion technically slows things down if you're on a _consistent_ connection, even if the connection is really slow. It adds quite a bit of chattiness over the network with lots of POST, PUT, and OPTIONS requests, which ironically would cause it to work poorly on an unstable connection (at least until it gets started).

If I understand how the intelligent commit works, you're basically combining the chunks yourself, thus adding additional steps and roundtrips to the entire upload process. This seems to imply that even in perfect network conditions, the intelligent upload would take longer than a normal multi-part one. It these assumptions are correct, the `intelligent: 'fallback'` option sounds like a perfect idea, and probably a good default.

However, I've been playing around with this in my dev environment, but it doesn't seem to work. With that on and a really small `intelligentChunkSize`, I can tell that there are tons of chunks getting uploaded at once in the Network inspector. I also see the start XHR request says `upload_type: "intelligent_ingestion"` and that the chunks are being uploaded to a `/ii/*` path.

Here are my settings:
```
        uploadConfig:
          intelligent: 'fallback'
          # Raise the default 2 min timeout so that slow connections don't keep retrying.
          # Since we're using `intelligent: 'fallback'` mode, we want to raise the timeout to ensure the success
          # of the first try of each part on a slow connection, so that intelligent chunking never even kicks in.
          # The default chunk size is 6 MB, which is approx 6000 kb, and with a timeout of 10 minutes, and with 3
          # concurrent connections, a client would need to upload at 30 kb / sec to get the chunks out
          # within the timeout. This is a reasonable assumption for most decent connections. Once any of the original
          # parts fails, the uploader will fallback to the smaller intelligent chunk size below.
          timeout: 10 * 60 * 1000
          # Since intelligent chunks only kick in after a normal chunk fails, we want them to be small (1 MB) so they
          # will be more likely to succeed on a slower or flakey connection.
          intelligentChunkSize: 1 * 1024 * 1024
          onRetry: (obj) ->
            console.debug "Retrying #{obj.location} for #{obj.filename}, attempt #{obj.attempt} of 10" +
            ", location #{obj.location}", obj
```

**Comment by velveteer** [comment]
Ah sorry, I forgot that the picker doesn't respect the `fallback` value yet -- it was added for another use case that was only using the raw `upload` method. That will be fixed in a patch next week, and we may as well just make it the default because you are right, the extra overhead of FII enabled by default will slow things down a little bit.

I will update this issue when that patch lands.

**Comment by wuservices** [comment]
Thanks for verifying on those things. I'll deploy with intelligent on but raise the default chunk size for now, but we're looking forward to that new release!

As a side note, it kind of seems like Google Cloud Storage supports chunks in multiples of 256 KB out of the box https://cloud.google.com/storage/docs/json_api/v1/how-tos/resumable-upload. I wonder if that's a lot simpler and more efficient than you guys handling it on your own in conjunction with S3.

**Comment by velveteer** [comment]
@wuservices 0.10.0 is now live, and it includes a patch that lets `intelligent: 'fallback'` be used in the picker. Feel free to open another issue if you run into problems with this release.
